src/cmt_decode_msgpack.c

```diff
@@ -0,0 +1,647 @@
+
+static struct cmt_map_label *find_label_by_index(struct mk_list *label_list, size_t desired_index)
+{
+    struct mk_list *head;
+    size_t          entry_index;
+    
+    entry_index = 0;
+
+    mk_list_foreach(head, label_list) {
+        if (entry_index == desired_index) {
+            return mk_list_entry(head, struct cmt_map_label, _head);
+        }
+
+        entry_index++;
+    }
+
+    return NULL;    
+}
+
+static int unpack_opts_ns(mpack_reader_t *reader, size_t index, void *context)
+{
+    struct cmt_opts *opts;
+
+    opts = (struct cmt_opts *) context;
+
+    return cmt_mpack_consume_string_tag(reader, &opts->namespace);
+}
+
+static int unpack_opts_ss(mpack_reader_t *reader, size_t index, void *context)
+{
+    struct cmt_opts *opts;
+    
+    opts = (struct cmt_opts *) context;
+
+    return cmt_mpack_consume_string_tag(reader, &opts->subsystem);
+}
+
+static int unpack_opts_name(mpack_reader_t *reader, size_t index, void *context)
+{
+    struct cmt_opts *opts;
+    
+    opts = (struct cmt_opts *) context;
+
+    return cmt_mpack_consume_string_tag(reader, &opts->name);
+}
+
+static int unpack_opts_desc(mpack_reader_t *reader, size_t index, void *context)
+{
+    struct cmt_opts *opts;
+    
+    opts = (struct cmt_opts *) context;
+
+    return cmt_mpack_consume_string_tag(reader, &opts->description);
+}
+
+static int unpack_opts(mpack_reader_t *reader, struct cmt_opts *opts)
+{
+    int                                   result;
+    struct cmt_mpack_map_entry_callback_t callbacks[] = {
+                                                            {"ns",     unpack_opts_ns},
+                                                            {"ss",     unpack_opts_ss},
+                                                            {"name",   unpack_opts_name},
+                                                            {"desc",   unpack_opts_desc},
+                                                            {NULL,     NULL}
+                                                        };
+
+    if (NULL == reader || 
+        NULL == opts   ) {
+        return CMT_DECODE_MSGPACK_INVALID_ARGUMENT_ERROR;
+    }
+
+    memset(opts, 0, sizeof(struct cmt_opts));
+
+    result = cmt_mpack_unpack_map(reader, callbacks, (void *) opts);
+
+    if (CMT_DECODE_MSGPACK_SUCCESS == result) {
+        /* Allocate enough space for the three components, the separators 
+         * and the terminator so we don't have to worry about possible realloc issues
+         * later on.
+         */
+
+        opts->fqname = cmt_sds_create_size(cmt_sds_len(opts->namespace) + \
+                                           cmt_sds_len(opts->subsystem) + \
+                                           cmt_sds_len(opts->name) + \
+                                           4); 
+
+        if (NULL == opts->fqname) {
+            return CMT_DECODE_MSGPACK_ALLOCATION_ERROR; 
+        }
+
+        cmt_sds_cat(opts->fqname, opts->namespace, cmt_sds_len(opts->namespace));
```

Comment:
please use cmt_sds_cat_safe, opts->fqname sds might be reallocated and the memory address be different

Reply:
There should be no way for it to realloc because I purposefully created the cmt_sds_t with the total required size (so I could check for an allocation error only once). However, I couldn't find cmt_sds_cat_safe in the source code.